### ISSUE TITLE
Track SparkSubmitHook Yarn Cluster application with Yarn CLI 

### ISSUE THREAD

**Issue by tokoko** [apache/airflow#24171]
### Description

SparkSubmitHook should track yarn cluster-mode application status with yarn CLI rather than rely on spark-submit process logs. This would cut back on excessive memory usage and also make it much easier to make the operator deferrable later on.

### Use case/motivation

While running most of our Spark workloads in Yarn cluster mode using SparkSubmitHook, we observed that celery workers were consistently low on memory. The main driver for the high memory consumption were spark-submit processes started from SparkSubmitHook, that took about 500mb of memory even though in yarn cluster mode they were doing essentially next to none of actual work. We refactored the hook to kill spark-submit process right after Yarn accepts the application and track the status with `yarn application -status` calls similar to how spark standalone mode is being tracked.

Another motivation for the change is to prepare the operator to be made deferrable later on. Polling from external java process that needs to be kept alive until operator exits can't be made deferrable. Using Yarn CLI for polling would fit easily with how deferrable operators work.

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by vchiapaikeo** [comment]
Happy to give this a shot!

**Comment by tokoko** [comment]
@vchiapaikeo Hey, I went through several iterations of this since this issue. The version I'm trying to adopt right now is deferrable and uses yarn rest api rather than yarn cli, because making `aiohttp` calls from triggerer is a lot simpler than launching subprocesses. I'm using a modified version of `HttpTrigger` (the one that can listen for specific sentinel values being returned as response content) to listen for application state changes. I can share the code later if that helps.

**Comment by vchiapaikeo** [comment]
That'd be great @tokoko - I'm familiar with turning http processes into deferrable. My concern here is that the [SparkSubmitHook](https://github.com/apache/airflow/blob/dcb72b5a4661223c9de7beea40264a152298f24b/airflow/providers/apache/spark/hooks/spark_submit.py#L39) supports both the client and cluster deploy modes and seemingly [3 cluster manager modes](https://github.com/apache/airflow/blob/dcb72b5a4661223c9de7beea40264a152298f24b/airflow/providers/apache/spark/hooks/spark_submit.py#L159-L160). This solution would only be applicable with Yarn deployment. I'd also likely develop locally with standalone mode as well so would not be able to make this rest call. 

Do you know of a comprehensive solution to [track driver status](https://github.com/apache/airflow/blob/dcb72b5a4661223c9de7beea40264a152298f24b/airflow/providers/apache/spark/hooks/spark_submit.py#L549) that would work for standalone, k8's, and yarn?

**Comment by tokoko** [comment]
@vchiapaikeo that's right. Full solution is really tricky here, that's why I chose to keep the implementation in-house, it was mostly just suited to our environment only :) Running `client` and `local` jobs in deferrable mode doesn't really make too much sense, I think. I would just throw an error in that case. 

For `cluster` mode jobs, there's no comprehensive solution AFAIK. You can track standalone jobs with `spark-submit` itself, but you can't do the same in case of yarn and k8s. For them you would either use respective clients (`yarn` and `kubectl`) or rest api calls. A web of if-else clauses is probably the only way to go here.

That's true, tracking YARN with rest calls assumes that rest api is exposed, which might or might not be the case depending on the environment. in case of k8s, it's probably safe to assume that both kubectl and rest api will be available.

**Comment by vchiapaikeo** [comment]
Ooo @tokoko , I just recently learned that Spark master exposes a rest API with `spark.master.rest.enabled true`. Can you check if that's the case on your yarn deployment? If possible to access on Yarn, maybe we can use this to obtain status?

I am running Spark on minikube following these [instructions](https://testdriven.io/blog/deploying-spark-on-kubernetes/).

<img width="793" alt="image" src="https://github.com/apache/airflow/assets/9200263/47914c03-ff91-4875-9c0f-ebc9eab70ec4">

Rest API: 

<img width="1323" alt="image" src="https://github.com/apache/airflow/assets/9200263/c3c51732-77f8-4252-891a-cb63a07c5a0a">

**Comment by tokoko** [comment]
Depends on what you mean exactly. YARN always exposes rest api, it just might or might not be accessible from airflow worker because of network security policies. The address can be configured as `yarn.resourcemanager.webapp.address` in `yarn-site.xml`.

So, all three cluster managers do expose rest apis, but the specifics of the apis are different. For example in YARN you have to call http://{yarn_address}/ws/v1/cluster/apps/{application_id}/state to get the necessary state. Take a look at this gist https://gist.github.com/tokoko/f559fb164c9433bb09c9618bf2949f33.

**Comment by tirkarthi** [comment]
I just want to add that we have been using yarn rest API to fetch application status in our custom deferrable operators related to spark on yarn and it has been working fine. We had also implemented the logic to kill the application when the task gets cleared or marked as success/fail from UI.

https://hadoop.apache.org/docs/stable/hadoop-yarn/hadoop-yarn-site/NodeManagerRest.html#Application_API

**Comment by potiuk** [comment]
@vchiapaikeo We are unassigning you from this issue as part of our updated [assignment policy](https://github.com/apache/airflow/blob/main/contributing-docs/04_how_to_contribute.rst#contribute-code-changes).

This is not meant to discourage your contribution — quite the opposite! You are still very welcome to work on this issue and submit a PR for it. Simply comment that you are working on it and open a PR when ready.

We found that formal assignments were not working well, as they often prevented others from contributing when the assignee was not actively working on the issue.

**Comment by kaxil** [comment]
cc @amoghrajesh 

Thanks for bringing it up @tirkarthi on the dev call

**Comment by amoghrajesh** [comment]
I had created a general ticket for this here: https://github.com/apache/airflow/issues/67168, does this relate well / cover it?

**Comment by amoghrajesh** [comment]
Yes, this is directly related. Quick summary of what I think can be done:

https://github.com/apache/airflow/pull/65991 solves the memory problem here, terminate the spark-submit early after YARN accepts the submission, poll via `yarn application -status`. That is the "non-blocking submit" split.

https://github.com/apache/airflow/issues/67168 will be intending to solve the crash recovery problem on top of that split — once `spark-submit` returns the app ID immediately, we persist it to `task_state` and reconnect on retry instead of resubmitting a duplicate job.

The two are complementary layers. #65991 is a prerequisite for my work in a sens, ie: it makes the hook return the app ID early, which is what we need to persist.

One coordination point worth discussing: #65991 uses `yarn application -status` (CLI subprocess) for polling. My plan was to use the YARN RM REST API (GET `/ws/v1/cluster/apps/{id}`). REST avoids spawning a subprocess and does not require yarn CLI on the worker, but it is worth aligning rather than having two different polling mechanisms in the same codebase.

Happy to sync if useful @nailo2c

**Comment by aaron-y-chen** [comment]
Hi @amoghrajesh, thanks for the summary. I think I can look into refactoring my part to use the REST API.

**Comment by amoghrajesh** [comment]
@nailo2c feel free to drop me a text on slack too, or join [#sig-aip-103-task-state-management](https://apache-airflow.slack.com/archives/C0AUE0S76QM) to discuss further


EDIT: I added you

**Comment by amoghrajesh** [comment]
Same problem exists for spark on k8s. The `spark-submit` blocks for the full job duration watching the driver pod via `LoggingPodStatusWatcherImpl`, holding similar heap size while doing no actual work. Also blocks making the operator deferrable when/if needed.

Planning to fix it with the same pattern as #65991 (yarn_track_via_rm_api): a `track_driver_via_k8s_api: bool = False` flag on `SparkSubmitHook/SparkSubmitOperator` that terminates `spark-submit` early after pod creation and polls via the K8s client instead.

**Comment by amoghrajesh** [comment]
Took a stab at it here: https://github.com/apache/airflow/pull/67715
